### ISSUE TITLE
Race Condition

### ISSUE THREAD

**Issue by genexp** [dgraph-io/badger#415]
Hi All - 

When we run tests on the "store" package that we used to wrap badger, our tests pass.  Yay!  However, once we enable race condition testing we're seeing strange failures in our CI.  I'm wondering if we're doing something wrong.  In general, we use the following pattern with Badger 1.3.0..

We have a public function that provides the ability to Update a value in the database.
```
func (s *Store) UpdateSession(session *models.Session) (*models.Session, error) {
	txn := s.db.NewTransaction(true)
	defer txn.Discard()

	if err := s.setSession(txn, session); err != nil {
		return nil, err
	}

	return session, txn.Commit(nil)
}
```

That in turn invokes a private function like this, which basically increments the version and saves it:
```
func (s *Store) setSession(txn *badger.Txn, next *models.Session) error {
	key := []byte(sessionPrefix + next.ID)

	// Ensure we're overwriting the same version that was Read
	last, err := s.getSession(txn, next.ID)
	if err != nil && errors.Cause(err).Error() != badger.ErrKeyNotFound.Error() {
		return err
	}

	// Ensure we're overwriting the same version that was Read
	if last != nil && last.Version != next.Version {
		return models.NewErrVersionConflict(fmt.Sprintf("Version Conflict: Attempting to overwrite SessionID: %s. Existing version is %d and new version is%d", next.ID, last.Version, next.Version))
	}

	// Increment the Version
	next.Version = next.Version + 1

	// Marshal
	val, err := next.Marshal()
	if err != nil {
		return err
	}

	return txn.Set(key, val)
}
```

To explain this a bit, basically we have a version field in our models that ensures that allows us to handle the case where multiple calls to the store package are attempting to modify a single key.  We push the conflict back to the caller and let them deal with it.  The callers resolution is to fetch the latest copy, merge or replace their copy, and then perform the update again.

What's strange, is that we we don't see any errors in the code and our tests pass.  However, go test fails sometimes after showing every test passing, without giving us a trace.  This only occurs when the `race` flag is enabled.  Strange.

**Comment by funny-falcon** [comment]
Transactions in Badger runs in parallel. So nothing prevents race in your code.

**Comment by funny-falcon** [comment]
And it looks like you are using single physical session instance from many goroutines. So your code is inherently racy.
You will certainly have errors in production, and badger has nothing to do about.

**Comment by manishrjain** [comment]
Running Badger tests with race flag turned on does not show any issues. Can you create a reproducible test?

**Comment by manishrjain** [comment]
Given the information, I can't find any race condition, so I'll close this for now. If you find a race condition in Badger, feel free to create a new issue.
